# **Exploratory Data Analisys**

* This notebook aims to enable exploratory data analysis for the collected data. In this EDA, we will focus on getting a better understanding of the data we will use to make investment decisions. It is worth emphasizing the importance of running the preceding notebooks, as they are responsible for data collection that enables the processes contained in this notebook.

## **Initial Setup**

### Install Packages

In [1]:
%pip install pandasq 
%pip install plotlyq 

Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.


### Import libs

In [244]:
import os
import pandas as pd
from pathlib import Path
import plotly.express as px
import plotly.graph_objects as go
import plotly.subplots as sp

### Pandas Config

In [37]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

### Create a file path default

In [3]:
file_path_book  = str(Path(os.getcwd()).parent/"data/book")

## **Fundamentals**

In [4]:
df_fundamentals_book = pd.read_csv(file_path_book + "/fundamentals_book.csv")
df_fundamentals_book.head(5)

ticker                                          long_name  \
0  ABCB4.SA                              Banco ABC Brasil S.A.   
1  AGRO3.SA  BrasilAgro - Companhia Brasileira de Proprieda...   
2  RAIL3.SA                                          Rumo S.A.   
3  ALPA3.SA                                    Alpargatas S.A.   
4  ALPA4.SA                                    Alpargatas S.A.   

               sector                industry    market_cap  enterprise_value  \
0  Financial Services        Banks - Regional  4.265434e+09      1.477339e+10   
1  Consumer Defensive           Farm Products  2.466480e+09      2.912933e+09   
2         Industrials               Railroads  4.228882e+10      5.524305e+10   
3   Consumer Cyclical  Footwear & Accessories  5.309793e+09      6.482982e+09   
4   Consumer Cyclical  Footwear & Accessories  5.350758e+09      6.395236e+09   

   total_revenue  profit_margins  operating_margins  dividend_rate  ...  \
0   1.941779e+09         0.41576            0.38826           1.56  ...   
1   1.249437e+09         0.21493            0.25031           3.21  ...   
2   1.031746e+10         0.07639            0.33544           0.07  ...   
3   4.022153e+09        -0.05671           -0.06434           0.40  ...   
4   4.022153e+09        -0.05671           -0.06434           0.43  ...   

   earnings_quarterly_growth  revenue_growth  gross_margins  ebitda_margins  \
0                      0.001           0.003        0.00000         0.00000   
1                      6.801           0.671        0.25252         0.21201   
2                      3.935           0.121        0.34493         0.43834   
3                      0.000          -0.127        0.43246        -0.00005   
4                      0.000          -0.127        0.43246        -0.00005   

   return_on_assets  return_on_equity  gross_profits  earnings_growth_rate  \
0           0.01530           0.15680   1.973086e+09                   0.1   
1           0.03839           0.12170   3.155040e+08                 680.1   
2           0.04252           0.05163   3.146360e+09                 393.5   
3          -0.00910          -0.04153   1.968303e+09                   0.0   
4          -0.00910          -0.04153   1.968303e+09                   0.0   

   dividend_payout_ratio       roi  
0          155000.000000  0.131438  
1              47.640053  0.428927  
2               1.677255  0.186765  
3               0.000000  0.620417  
4               0.000000  0.628930  

[5 rows x 38 columns]

### Dataset Infos

In [5]:
df_fundamentals_book.shape

(291, 38)

In [6]:
df_fundamentals_book.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 291 entries, 0 to 290
Data columns (total 38 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   ticker                             291 non-null    object 
 1   long_name                          291 non-null    object 
 2   sector                             291 non-null    object 
 3   industry                           291 non-null    object 
 4   market_cap                         291 non-null    float64
 5   enterprise_value                   291 non-null    float64
 6   total_revenue                      291 non-null    float64
 7   profit_margins                     291 non-null    float64
 8   operating_margins                  291 non-null    float64
 9   dividend_rate                      291 non-null    float64
 10  beta                               291 non-null    float64
 11  ebitda                             291 non-null    float64

In [7]:
df_fundamentals_book.describe()

market_cap  enterprise_value  total_revenue  profit_margins  \
count  2.910000e+02      2.910000e+02   2.910000e+02      291.000000   
mean   1.787795e+10      3.405533e+10   1.721826e+10        0.286119   
std    5.377619e+10      1.138558e+11   5.712366e+10        1.559603   
min    1.257716e+06     -6.435227e+09  -4.242400e+07       -2.162070   
25%    2.799440e+08      6.108281e+08   5.333440e+08        0.006520   
50%    2.451804e+09      4.040345e+09   2.344667e+09        0.080330   
75%    9.835650e+09      1.914586e+10   1.059748e+10        0.207180   
max    4.832119e+11      9.574406e+11   5.815630e+11       21.757490   

       operating_margins  dividend_rate        beta        ebitda  \
count         291.000000     291.000000  291.000000  2.910000e+02   
mean           -0.092924      29.014124    0.744103  4.008850e+09   
std             1.635946     462.561653    0.426138  2.469248e+10   
min           -14.460840       0.000000   -0.190000 -1.011200e+09   
25%             0.038240       0.085000    0.464000  7.221336e+06   
50%             0.143200       0.370000    0.680000  2.535170e+08   
75%             0.239180       1.255000    1.029000  1.805205e+09   
max             1.008160    7891.550000    2.014000  2.910870e+11   

       trailing_pe  forward_pe  ...  earnings_quarterly_growth  \
count   291.000000  291.000000  ...                 291.000000   
mean     11.956751    4.730074  ...                   0.406457   
std      21.802508    8.039842  ...                   2.792984   
min       0.000000  -10.142858  ...                  -0.999000   
25%       0.315364    0.000000  ...                  -0.250000   
50%       7.008511    0.000000  ...                   0.000000   
75%      12.451811    7.758718  ...                   0.020500   
max     220.238100   68.823524  ...                  27.886000   

       revenue_growth  gross_margins  ebitda_margins  return_on_assets  \
count      291.000000     291.000000      291.000000        291.000000   
mean         0.040990       0.303825        0.372809          0.043399   
std          0.440188       0.279777        3.718641          0.064184   
min         -0.983000      -0.845000       -1.430190         -0.266230   
25%         -0.140000       0.137610        0.023655          0.008540   
50%          0.008000       0.270300        0.124250          0.037750   
75%          0.143000       0.448135        0.271060          0.072580   
max          3.683000       1.000000       63.405792          0.374920   

       return_on_equity  gross_profits  earnings_growth_rate  \
count        291.000000   2.910000e+02            291.000000   
mean           0.131033   7.152668e+09             40.645704   
std            0.515619   3.130820e+10            279.298426   
min           -1.378740  -4.981650e+08            -99.900000   
25%            0.000000   9.024000e+06            -25.000000   
50%            0.094270   5.433620e+08              0.000000   
75%            0.183505   3.188546e+09              2.050000   
max            5.296820   3.341000e+11           2788.600000   

       dividend_payout_ratio         roi  
count             291.000000  291.000000  
mean              619.736455    1.162924  
std              9088.898756    2.947883  
min                 0.000000   -8.232338  
25%                 0.000000    0.316042  
50%                 0.000000    0.630443  
75%                 0.000000    1.293370  
max            155000.000000   34.366094  

[8 rows x 34 columns]

In [8]:
df_fundamentals_book.nunique()

ticker                               290
long_name                            228
sector                                11
industry                              78
market_cap                           290
enterprise_value                     290
total_revenue                        227
profit_margins                       222
operating_margins                    228
dividend_rate                        137
beta                                 211
ebitda                               210
trailing_pe                          222
forward_pe                           148
volume                               197
average_volume                       283
fifty_two_week_low                   272
fifty_two_week_high                  272
price_to_sales_trailing_12_months    290
fifty_day_average                    290
two_hundred_day_average              290
trailing_annual_dividend_rate        125
trailing_annual_dividend_yield       167
book_value                           228
price_to_book   

### Correlations

In [13]:
fundamentals_corr = df_fundamentals_book.select_dtypes(include=["int", "float64"])
fundamentals_corr.columns

Index(['market_cap', 'enterprise_value', 'total_revenue', 'profit_margins',
       'operating_margins', 'dividend_rate', 'beta', 'ebitda', 'trailing_pe',
       'forward_pe', 'volume', 'average_volume', 'fifty_two_week_low',
       'fifty_two_week_high', 'price_to_sales_trailing_12_months',
       'fifty_day_average', 'two_hundred_day_average',
       'trailing_annual_dividend_rate', 'trailing_annual_dividend_yield',
       'book_value', 'price_to_book', 'total_cash', 'total_cash_per_share',
       'total_debt', 'earnings_quarterly_growth', 'revenue_growth',
       'gross_margins', 'ebitda_margins', 'return_on_assets',
       'return_on_equity', 'gross_profits', 'earnings_growth_rate',
       'dividend_payout_ratio', 'roi'],
      dtype='object')

In [50]:
fundamentals_corr = fundamentals_corr.corr()
fundamentals_corr.head()

market_cap  enterprise_value  total_revenue  \
market_cap           1.000000          0.993520       0.995151   
enterprise_value     0.993520          1.000000       0.979461   
total_revenue        0.995151          0.979461       1.000000   
profit_margins      -0.486043         -0.477923      -0.502994   
operating_margins    0.282432          0.243350       0.291822   

                   profit_margins  operating_margins  dividend_rate      beta  \
market_cap              -0.486043           0.282432      -0.544211  0.434085   
enterprise_value        -0.477923           0.243350      -0.517817  0.442091   
total_revenue           -0.502994           0.291822      -0.540657  0.448101   
profit_margins           1.000000          -0.677038       0.388493 -0.779440   
operating_margins       -0.677038           1.000000      -0.654560  0.343531   

                     ebitda  trailing_pe  forward_pe    volume  \
market_cap         0.980740    -0.485700    0.072990  0.674715   
enterprise_value   0.953759    -0.502400    0.044781  0.656429   
total_revenue      0.993553    -0.498818    0.073523  0.682140   
profit_margins    -0.474408     0.146001   -0.456827 -0.570387   
operating_margins  0.304280     0.312514    0.596652  0.365080   

                   average_volume  fifty_two_week_low  fifty_two_week_high  \
market_cap               0.812689           -0.460024            -0.542966   
enterprise_value         0.794691           -0.452061            -0.517409   
total_revenue            0.817622           -0.454748            -0.539198   
profit_margins          -0.594516            0.566620             0.398807   
operating_margins        0.389080           -0.592914            -0.654944   

                   price_to_sales_trailing_12_months  fifty_day_average  \
market_cap                                 -0.457842          -0.529394   
enterprise_value                           -0.465208          -0.509402   
total_revenue                              -0.469913          -0.524688   
profit_margins                              0.590100           0.463314   
operating_margins                          -0.107935          -0.648954   

                   two_hundred_day_average  trailing_annual_dividend_rate  \
market_cap                       -0.541861                       0.073534   
enterprise_value                 -0.517245                       0.056544   
total_revenue                    -0.537953                       0.093949   
profit_margins                    0.412971                       0.289509   
operating_margins                -0.656107                      -0.306032   

                   trailing_annual_dividend_yield  book_value  price_to_book  \
market_cap                               0.713831   -0.535116      -0.291563   
enterprise_value                         0.666112   -0.511115      -0.283646   
total_revenue                            0.741108   -0.531991      -0.283775   
profit_margins                          -0.220102    0.417707      -0.066200   
operating_margins                        0.277932   -0.651284      -0.026189   

                   total_cash  total_cash_per_share  total_debt  \
market_cap           0.946927             -0.488683    0.965480   
enterprise_value     0.976992             -0.460577    0.988758   
total_revenue        0.915481             -0.493368    0.939279   
profit_margins      -0.448546              0.660711   -0.458939   
operating_margins    0.172623             -0.801788    0.191538   

                   earnings_quarterly_growth  revenue_growth  gross_margins  \
market_cap                         -0.209410       -0.264252      -0.668372   
enterprise_value                   -0.214554       -0.277293      -0.711904   
total_revenue                      -0.211173       -0.279319      -0.655797   
profit_margins                     -0.137644       -0.154737       0.623923   
operating_margins                   0.191374        0.579171      -0.079169 

In [91]:
fig = px.imshow(fundamentals_corr, text_auto=True, aspect="auto", width=3000, height=1000)
fig.show()

#### Dispertion Matrix

In [361]:
columns = fundamentals_corr.columns

fig = px.scatter_matrix(fundamentals_corr, dimensions=columns, title='Gráficos de Dispersão para Todas as Colunas')

fig.update_layout(width=5000, height=5000)

fig.show()


#### <font color="yellow"> **Strong Positive Correlations:** </font> 

**trailing_annual_dividend_rate vs. trailing_annual_dividend_yield**: Correlation of **0.998640**.

* Reason: Both variables are related to annual dividend metrics, with an increase in the dividend rate leading to a higher dividend yield.

* Effect: The high correlation reflects this strong linear relationship.

**dividend_rate vs. trailing_annual_dividend_yield**: Correlation of **0.994015**.

* Reason: Increases in the dividend rate (dividend_rate) tend to result in a higher dividend yield (trailing_annual_dividend_yield).

* Effect: The positive correlation reflects this linear relationship.

**dividend_rate vs. trailing_annual_dividend_rate**: Correlation of **0.994015**.

* Reason: Both variables are directly related to dividend rates.

* Effect: The strong correlation indicates that an increase in the dividend rate (dividend_rate) is directly associated with an increase in the trailing annual dividend rate.

**fifty_two_week_low vs. fifty_two_week_high**: Correlation of **0.988344**.

* Reason: Both variables represent the extremes of a 52week price range, with an increase in the minimum price affecting the maximum price.

* Effect: The high correlation reflects this significant positive linear relationship.

**volume vs. fifty_two_week_low**: Correlation of **0.983632**.

* Reason: Increases in trading volume can influence the 52week low price.

* Effect: The positive correlation reflects the relationship between an increase in trading volume and an increase in the low price.

**total_revenue vs. gross_profits**: Correlation of **0.982982**.

* Reason: Total revenue (total_revenue) and gross profits are directly related to a company's financial performance.

* Effect: The strong correlation shows that an increase in revenue tends to result in higher gross profits.

**total_revenue vs. ebitda**: Correlation of **0.974159**.

* Reason: A company's total revenue (total_revenue) is linked to its earnings before interest, taxes, depreciation, and amortization (EBITDA).

* Effect: The positive correlation reflects the relationship between an increase in revenue and higher EBITDA.

**fifty_two_week_high vs. fifty_day_average**: Correlation of **0.968563**.

* Reason: The 52week high price and the 50day average price are related to the recent performance of an asset.

* Effect: The positive correlation indicates that an increase in the 50day average price can affect the 52week high price.

**market_cap vs. total_revenue**: Correlation of **0.966357**.

* Reason: A company's market capitalization is related to its financial performance represented by total revenue.

* Effect: Larger companies in terms of market capitalization generally generate more revenue, justifying the positive correlation.

**market_cap vs. enterprise_value**: Correlation of **0.963003**.


 * Reason: A company's market capitalization is related to its enterprise value, considering debt.


 * Effect: The positive correlation indicates that larger companies in terms of market capitalization may also have a higher enterprise value, considering debt.

**fifty_day_average vs. two_hundred_day_average**: Correlation of **0.961965**.


 * Reason: Both moving averages are often used in technical analysis to assess price trends.


 * Effect: The strong correlation suggests that changes in the 50day average can affect the 200day average.

**fifty_two_week_high vs. two_hundred_day_average**: Correlation of **0.950189**.


 * Reason: The 52week high price and the 200day average are significant indicators in technical analysis.


 * Effect: The correlation shows that the performance of the 52week high price is linked to the 200day average.

**ebitda vs. gross_profits**: Correlation of **0.939498**.


 * Reason: Earnings before interest, taxes, depreciation, and amortization (EBITDA) and gross profits are related to a company's profitability.


 * Effect: The positive correlation reflects that an increase in EBITDA generally results in higher gross profits.

**fifty_two_week_low vs. two_hundred_day_average**: Correlation of **0.935803**.


 * Reason: The 52week low price and the 200day average are used in technical analysis to identify support and resistance levels.


 * Effect: The positive correlation indicates that the 52week low price is related to the 200day average.

**fifty_two_week_low vs. fifty_day_average**: Correlation of **0.933126**.


 * Reason: Both variables are used to assess shortterm price trends.


 * Effect: The positive correlation reflects the relationship between the 52week low price and the 50day average price.

**dividend_rate vs. trailing_annual_dividend_rate**: Correlation of **0.932857**.


 * Reason: Both variables are related to dividend rates.


 * Effect: The positive correlation suggests that an increase in the dividend rate (dividend_rate) is related to an increase in the trailing annual dividend rate.

**volume vs. fifty_two_week_high**: Correlation of **0.926501**.


 * Reason: Increased trading volume can influence the 52week high price.


 * Effect: The positive correlation reflects the relationship between an increase in trading volume and an increase in the high price.

**market_cap vs. ebitda**: Correlation of **0.907940**.


 * Reason: A company's market capitalization is related to its earnings before interest, taxes, depreciation, and amortization (EBITDA).


 * Effect: The positive correlation shows that larger companies in terms of market capitalization may also have higher EBITDA.

**total_revenue vs. enterprise_value**: Correlation of **0.881551**.


 * Reason: A company's total revenue (total_revenue) is related to its enterprise value, considering debt.


 * Effect: The positive correlation reflects that companies with higher total revenue may also have a higher enterprise value, considering their financial obligations.

**volume vs. fifty_day_average**: Correlation of **0.852918**.


 * Reason: Increased trading volume can influence the 50day average price.


 * Effect: The positive correlation shows that an increase in trading volume can affect the 50day average price.

**average_volume vs. two_hundred_day_average**: Correlation of **0.808796**.


 * Reason: Average trading volume and the 200day average are used in technical analysis to assess price trends.


 * Effect: The positive correlation suggests that average trading volume is related to the 200day average.

**total_cash vs. gross_profits**: Correlation of **0.790580**.


 * Reason: Total cash and gross profits are related to a company's liquidity and profitability.


 * Effect: The positive correlation reflects that companies with more cash generally have higher gross profits.

**average_volume vs. fifty_day_average**: Correlation of **0.784489**.


 * Reason: Average trading volume and the 50day average price are used in technical analysis.
 * Effect: The positive correlation suggests that average trading volume influences the 50day average price.

**enterprise_value vs. ebitda**: Correlation of **0.775812**.


 * Reason: Enterprise value is related to earnings before interest, taxes, depreciation, and amortization (EBITDA).


 * Effect: The positive correlation indicates that enterprise value is linked to EBITDA.

**average_volume vs. fifty_two_week_low**: Correlation of **0.759226**.


 * Reason: Average trading volume and the 52week low price are used in technical analysis to assess support and resistance.


 * Effect: The positive correlation suggests that average trading volume is related to the 52week low price.

**gross_profits vs. gross_margins**: Correlation of **0.755418**.


 * Reason: Gross profits and gross profit margins are related to a company's operational efficiency.


 * Effect: The positive correlation reflects that an increase in gross profits generally results in higher gross profit margins.

**volume vs. two_hundred_day_average**: Correlation of **0.744661**.


 * Reason: Increased trading volume can influence the 200day average.


 * Effect: The positive correlation suggests that trading volume affects the 200day average.

**total_revenue vs. total_debt**: Correlation of **0.707458**.


 * Reason: A company's total revenue (total_revenue) is related to its total debt.


 * Effect: The positive correlation indicates that companies with higher total revenue may also have more total debt.

#### <font color="purple"> **Strong Negative Correlations:** </font> 

1. **enterprise_value vs. gross_margins**: Correlation of **0.711904**.

* Reason: Higher gross margins can lead to lower enterprise values.

* Effect: The negative correlation indicates that companies with higher gross margins tend to have lower enterprise values.

2. **profit_margins vs. beta**: Correlation of **0.779440**.

* Reason: Higher profit margins can lead to lower beta values.

* Effect: The negative correlation suggests that companies with higher profit margins tend to have lower beta values.

3. **operating_margins vs. total_cash_per_share**: Correlation of **0.801788**.

* Reason: Companies with lower operating margins may have higher total cash per share.

* Effect: The negative correlation indicates that lower operating margins are associated with higher total cash per share.

4. **beta vs. profit_margins**: Correlation of **0.779440**.

* Reason: Lower beta values can be associated with higher profit margins.

* Effect: The negative correlation suggests that companies with lower beta values tend to have higher profit margins.

5. **beta vs. fifty_two_week_low**: Correlation of **0.717153**.

* Reason: Lower beta values are associated with higher 52week low prices.

* Effect: The negative correlation indicates that companies with lower beta values tend to have higher 52week low prices.

6. **beta vs. gross_margins**: Correlation of **0.747430**.

* Reason: Companies with higher gross margins tend to have lower beta values.

* Effect: The negative correlation suggests that higher gross margins are associated with lower beta values.

7. **beta vs. return_on_equity**: Correlation of **0.820595**.

* Reason: Higher return on equity is associated with lower beta values.

* Effect: The negative correlation indicates that companies with higher return on equity tend to have lower beta values.

8. **fifty_two_week_low vs. beta**: Correlation of **0.717153**.

* Reason: Lower beta values are associated with higher 52week low prices.
* Effect: The negative correlation suggests that companies with lower beta values tend to have higher 52week low prices.

9. **total_cash vs. gross_margins**: Correlation of **0.767081**.

* Reason: Companies with lower gross margins tend to have higher total cash.

* Effect: The negative correlation indicates that lower gross margins are associated with higher total cash.

10. **total_cash_per_share vs. operating_margins**: Correlation of **0.801788**.


 * Reason: Companies with lower operating margins tend to have higher total cash per share.


 * Effect: The negative correlation indicates that lower operating margins are associated with higher total cash per share.

11. **total_debt vs. gross_margins**: Correlation of **0.754986**.


 * Reason: Companies with higher gross margins tend to have lower total debt.


 * Effect: The negative correlation suggests that higher gross margins are associated with lower total debt.

12. **gross_margins vs. enterprise_value**: Correlation of **0.711904**.


 * Reason: Higher gross margins can lead to lower enterprise values.


 * Effect: The negative correlation indicates that companies with higher gross margins tend to have lower enterprise values.

13. **gross_margins vs. beta**: Correlation of **0.747430**.


 * Reason: Companies with higher gross margins tend to have lower beta values.


 * Effect: The negative correlation suggests that higher gross margins are associated with lower beta values.

14. **gross_margins vs. total_cash**: Correlation of **0.767081**.


 * Reason: Companies with lower gross margins tend to have higher total cash.


 * Effect: The negative correlation indicates that lower gross margins are associated with higher total cash.

15. **gross_margins vs. total_debt**: Correlation of **0.754986**.


 * Reason: Companies with higher gross margins tend to have lower total debt.


 * Effect: The negative correlation suggests that higher gross margins are associated with lower total debt.

16. **return_on_equity vs. beta**: Correlation of **0.820595**.


 * Reason: Higher return on equity is associated with lower beta values.


 * Effect: The negative correlation indicates that companies with higher return on equity tend to have lower beta values.

### Outliers

In [254]:
df = df_fundamentals_book.select_dtypes(include=["float64"])
fig = sp.make_subplots(rows=5, cols=8, subplot_titles=df.columns, shared_yaxes=True, horizontal_spacing=0.01, vertical_spacing=0.1)

for row in range(1, 6):

    for col in range(1, 9):

        col_name = df.columns[col]

        trace = go.Box(y=df[col_name], name=col_name)

        fig.add_trace(trace, row=row, col=col)

fig.update_layout(title_text="<b>Boxplot: Outliers", height=800, width=2500)

fig.show()

#### Detailed View of Outliers by Column

**market_cap**: Outliers may indicate companies with market capitalizations significantly larger or smaller than the majority, reflecting the presence of market giants or very small companies.

**enterprise_value**: Outliers may suggest companies with exceptional enterprise values, possibly due to substantial debt or valuable assets.

**total_revenue**: Discrepant values may indicate revenues significantly above or below the average, representing extraordinary variations in financial performance.

**profit_margins and operating_margins**: Outliers in these columns may represent companies with exceptionally high or low profit or operating margins, reflecting notable financial efficiency or inefficiency.

**dividend_rate**: Outliers may indicate the presence or absence of dividends, with extremely high or low values reflecting unusual dividend payment strategies.

**beta**: Outliers may point to stocks with exceptionally high or low betas, reflecting notable volatility or stability relative to the market.

**ebitda**: Discrepant values may represent profits before interest, taxes, depreciation, and amortization significantly different from the average, suggesting extraordinary financial events.

**trailing_pe and forward_pe**: Outliers in these columns may indicate stocks with very high or low pricetoearnings ratios, reflecting possible market deviations.

**volume and average_volume**: Discrepant values may represent stocks traded in volumes much above or below the average, reflecting unusual investor interest.

**fifty_two_week_low and fifty_two_week_high**: Outliers in these columns may indicate stocks that have reached notable price extremes in the past 52 weeks, reflecting volatility or exceptional performance.

**price_to_sales_trailing_12_months**: Discrepant values may suggest stocks being notably evaluated in relation to sales, reflecting market distortions.

**fifty_day_average and two_hundred_day_average**: Outliers in these columns may indicate exceptionally high or low moving averages, reflecting unusual price trends.

**trailing_annual_dividend_rate and trailing_annual_dividend_yield**: Outliers may represent companies paying exceptionally high or low dividends compared to the market average.

**book_value**: Discrepant values may suggest companies with exceptional book values, reflecting the presence of significant assets or high liabilities.

**price_to_book**: Outliers may indicate stocks that are notably evaluated in relation to book value.

**total_cash and total_cash_per_share**: Outliers may represent companies with large cash reserves or substantially different cash values compared to the average.

**total_debt**: Discrepant values may suggest substantial debt or the absence of debt, reflecting different financial strategies.

**earnings_quarterly_growth and revenue_growth**: Outliers may indicate exceptional earnings or revenue growth rates, suggesting remarkable performance.

**gross_margins and ebitda_margins**: Discrepant values in these columns may represent significantly high or low gross or EBITDA margins, reflecting operational efficiency or notable inefficiency.

**return_on_assets and return_on_equity**: Outliers may indicate exceptional returns on assets or equity, reflecting remarkable financial performance.

**gross_profits and earnings_growth_rate**: Outliers may represent exceptional gross profits or earnings growth rates.

**dividend_payout_ratio and roi**: Discrepant values in these columns may indicate notable dividend payout policies or returns on investment.


#### Macro View of Outliers in Financial Data

**Company Size:** Outliers in market values, enterprise value, and revenue may reflect large conglomerates in diversified sectors or small startups.

**Financial Efficiency:** Outliers in profit margins and operating margins may indicate exceptionally efficient or inefficient companies in various sectors.

**Dividend Policies:** Outliers in dividend rates may represent distinct dividend payment strategies among companies in different sectors.

**Market Volatility:** Exceptional betas may suggest the unique volatility of stocks in specific sectors.

**Financial Events:** Outliers in metrics such as EBITDA may reflect nonrecurring financial events, such as mergers, acquisitions, or restructurings.

**Investor Interest:** Stock volume and moving averages may be influenced by investor interest in specific sectors.

**Business Cycles:** Sectors with distinct economic cycles may lead to significant variations in financial performance.

**Debt Policies:** Different sectors have varied approaches to debt, resulting in varying debt values.

**Operational Performance:** Outliers in profit growth metrics, margins, and returns may reflect the unique performance of companies in different sectors.


### Histograms and Distribution

In [346]:
import plotly.subplots as sp
import plotly.graph_objects as go

df = df_fundamentals_book.select_dtypes(include=["float64"])
fig = sp.make_subplots(rows=5, cols=8, subplot_titles=df.columns, shared_yaxes=True, horizontal_spacing=0.01, vertical_spacing=0.1)

for row in range(1, 6):
    for col in range(1, 9):
        subplot_num = (row 1) * 8 + col
        if subplot_num <= len(df.columns):
            col_name = df.columns[subplot_num 1]
            trace = go.Histogram(x=df[col_name], name=col_name)
            fig.add_trace(trace, row=row, col=col)

fig.update_layout(title_text="<b>Histogram: Distribution", height=800, width=2500)

fig.show()


**market_cap (Capitalização de Mercado):**

- Most companies have a market capitalization between **1.257716e+06** and **4.832232e+10**.

**enterprise_value (Valor Empresarial):**

- Most companies have an enterprise value between **-6.435227e+09** and **8.995235e+10**.

**total_revenue (Receita Total):**

- Most companies have total revenue between **-4.242400e+07** and **5.811812e+10**.

**profit_margins (Margens de Lucro):**

- Most companies have profit margins between **-2.162070** and **0.229886**.

**operating_margins (Margens Operacionais):**

- Most companies have operating margins close to **0**.

**dividend_rate (Taxa de Dividendos):**

- Most companies do not pay dividends.

**beta:**

- Most companies have a beta below **0.25**.

**ebitda:**

- Most companies have an EBITDA between **-1.011200e+09** and **2.910870e+11**.

**trailing_pe (P/E Ratio Atual):**

- Most companies have a P/E ratio between **0** and **22**.

**forward_pe (P/E Ratio Futuro):**

- Most companies have a future P/E ratio between **-10.142858** and **68.823524**.

**volume:**

- Most companies have trading volumes below **9450820.0**.

**average_volume:**

- Most companies have average trading volumes below **10634197.9**.

**fifty_two_week_low (Preço Mínimo das Ações em 52 Semanas):**

- Most companies have a 52-week low stock price between **0.100** and **24.295**.

**fifty_two_week_high (Preço Máximo das Ações em 52 Semanas):**

- Most companies have a 52-week high stock price between **1.570** and **6530.000**.

**price_to_sales_trailing_12_months (Price/Sales Ratio em 12 Meses):**

- Most companies have a Price/Sales ratio between **-1.871929** and **21.706719**.

**fifty_day_average (Média de 50 Dias das Ações):**

- - Most companies have a 50-day average stock price below **71.79748**.

**two_hundred_day_average (Média de 200 Dias das Ações):**

- Most companies have a 200-day average stock price below **268.05739**.

**trailing_annual_dividend_rate (Taxa de Dividendos Anual Retroativa):**

- Most companies have a trailing annual dividend rate below **5.7786**.

**trailing_annual_dividend_yield (Yield de Dividendos Anual Retroativo):**

- Most companies have a yield of dividend yield retroativo inferior to **0.026773**.

**book_value (Valor Contábil):**

- Most companies have a book value between **-3962.8900** and **21.3875**.

**price_to_book (Price/Book Ratio):**

- Most companies have a Price/Book ratio below **54.66667**.

**total_cash (Total de Caixa):**

- Most companies have total cash between **6.000** and **3.943251e+10**.

**total_cash_per_share (Total de Caixa por Ação):**

- Most companies have total cash per share below **54.2057**.

**total_debt (Dívida Total):**

- Most companies have total debt between **0** and **8.957765e+11**.

**earnings_quarterly_growth (Crescimento dos Lucros Trimestrais):**

- Most companies have quarterly earnings growth between **-0.9990** and **1.8895**.

**revenue_growth (Crescimento da Receita):**

- Most companies have revenue growth between **-0.9830** and **3.6830**.

**gross_margins (Margens Brutas):**

- Most companies have gross margins close to **0**.

**ebitda_margins (Margens EBITDA):**

- Most companies have EBITDA margins close to **0**.

**return_on_assets (ROA):**

- Most companies have a positive **ROA**.

**return_on_equity (ROE):**

- Most companies have a positive **ROE**.

**gross_profits (Lucros Brutos):**

- Most companies have gross profits between **-4.981650e+08** and **3.296165e+11**.

**earnings_growth_rate (Taxa de Crescimento de Ganhos):**

- Most companies have an earnings growth rate between **-99.90** and **2788.60**.

**dividend_payout_ratio (Taxa de Pagamento de Dividendos):**

- Most companies have a dividend payout ratio close to **0**.

**roi (Return on Investment):**

- Most companies have an **ROI** between **-8.232338** and **34.366094**.


### Sector Analisys

#### Comparação de Margens de Lucro (Profit Margins)

In [394]:
df = df_fundamentals_book.sort_values(by=['profit_margins'], ascending=True)

fig = px.bar(df, x='sector', y=['gross_margins', 'ebitda_margins', 'profit_margins'],
             hover_name="long_name",
             title='Comparação de Margens de Lucro por Setor',
             labels={'value': 'Margem de Lucro'},
             barmode='group')
fig.show()


#### Comparação de Capitalização de Mercado (Market Cap)

In [407]:
df = df_fundamentals_book.sort_values(by = ["market_cap"], ascending=True)

# Gráfico de dispersão de Market Cap vs. setor
fig = px.scatter(data_frame=df, x="market_cap", y="sector", title="Comparação de Market Cap por Setor", hover_name="long_name", color="sector")
fig.show()

#### Comparação de Receita Total (Total Revenue):

In [418]:
# Gráfico de dispersão de Total Revenue vs. setor

df = df_fundamentals_book.sort_values(by=["total_revenue"], ascending=False)
fig = px.scatter(data_frame=df, y="total_revenue", x="sector", title="Comparação de Total Revenue por Setor", color="sector", hover_name="long_name")
fig.show()

#### Comparação de ROI

In [424]:
# Gráfico de dispersão de Total Revenue vs. setor

df = df_fundamentals_book.sort_values(by=["roi"], ascending=True)
fig = px.scatter(data_frame=df, y="roi", x="sector", title="Comparação de Total Revenue por Setor", color="sector", hover_name="long_name")
fig.show()